server/src/main/java/org/elasticsearch/cluster/service/MasterService.java

```diff
@@ -1236,4 +1235,459 @@ synchronized ClusterStateUpdateStats getStatistics() {
+                        assert false : e2;
+                    }
+                }
+            }
+        }
+
+        private class Processor implements Batch {
+            @Override
+            public void onRejection(FailedToCommitClusterStateException e) {
+                final var items = queueSize.getAndSet(0);
+                for (int i = 0; i < items; i++) {
+                    final var entry = queue.poll();
+                    assert entry != null;
+                    entry.onRejection(e);
+                }
+            }
+
+            @Override
+            public void run() {
+                assert executing.isEmpty() : executing;
+                final var entryCount = queueSize.getAndSet(0);
+                var taskCount = 0;
+                for (int i = 0; i < entryCount; i++) {
+                    final var entry = queue.poll();
+                    assert entry != null;
+                    if (entry.acquireForExecution()) {
+                        taskCount += 1;
+                        executing.add(entry);
+                    }
+                }
+                if (taskCount == 0) {
+                    return;
+                }
+                final var finalTaskCount = taskCount;
+                final var tasks = new ArrayList<ExecutionResult<T>>(finalTaskCount);
+                for (final var entry : executing) {
+                    tasks.add(
+                        new ExecutionResult<>(entry.source(), entry.task(), threadPool.getThreadContext(), entry.storedContextSupplier())
+                    );
+                }
+                try {
+                    batchConsumer.runBatch(executor, tasks, new BatchSummary(() -> buildTasksDescription(tasks)));
+                } finally {
+                    assert executing.size() == finalTaskCount;
+                    executing.clear();
+                }
+            }
+
+            private String buildTasksDescription(List<ExecutionResult<T>> tasks) {
+                final var tasksBySource = new HashMap<String, List<T>>();
+                for (final var entry : tasks) {
+                    tasksBySource.computeIfAbsent(entry.getSource(), ignored -> new ArrayList<>()).add(entry.getTask());
+                }
+
+                final var output = new StringBuilder();
+                Strings.collectionToDelimitedStringWithLimit((Iterable<String>) () -> tasksBySource.entrySet().stream().map(entry -> {
+                    var tasksDescription = executor.describeTasks(entry.getValue());
+                    return tasksDescription.isEmpty() ? entry.getKey() : entry.getKey() + "[" + tasksDescription + "]";
+                }).filter(s -> s.isEmpty() == false).iterator(), ", ", "", "", MAX_TASK_DESCRIPTION_CHARS, output);
+                if (output.length() > MAX_TASK_DESCRIPTION_CHARS) {
+                    output.append(" (").append(tasks.size()).append(" tasks in total)");
+                }
```

Comment:
NIT: I think `collectionToDelimitedStringWithLimit` does this summary already, am I missing something?

Reply:
`collectionToDelimitedStringWithLimit` is iterating over the tasks grouped by source, so it'll report on the number of unique source strings rather than the total number of tasks.